src/freenas/usr/local/lib/middlewared_truenas/plugins/failover.py

```diff
@@ -1623,8 +1623,31 @@ async def service_remote(middleware, service, verb, options):
 
 
 async def ready_system_sync_keys(middleware):
+
     if await middleware.call('failover.licensed'):
-        await middleware.call('failover.call_remote', 'failover.sync_keys_to_remote_node')
+
+        # when middlewared is setup, we only run this if we're the standby controller
+        if not await middleware.call('failover.is_backup_node'):
+            return
+
+        remote_version = (await middleware.call('failover.call_remote', 'system.info'))['version'].split('-')
+        if len(remote_version) > 1 and remote_version[1].startswith('11'):
+            # remote system is running an older release (this happens on upgrades)
+            # use the legacy encryption methods
+            encrypted_pools = await middleware.call('pool.query', [('encrypt', '>=', 1)])
+            if encrypted_pools:
+                passphrase = await middleware.call('failover.call_remote', 'failover.encryption_getkey')
+                await middleware.call(
+                    'failover.update_encryption_keys', {
+                        'pools': [
+                            {'name': p['name'], 'passphrase': passphrase or ''}
+                            for p in encrypted_pools
+                        ],
+                        'sync_keys': False,
+                    }
+                )
+        else:
+            await middleware.call('failover.call_remote', 'failover.sync_keys_to_remote_node')
```

Comment:
If I am seeing correctly there is quite a big block of code identical here and in the event that could be "deduplicated"?

Reply:
yes, completely so. I'll fix.